src/CBLDatabase_Internal.hh

```diff
@@ -382,6 +390,46 @@ private:
     }
 
     void callDocListeners();
+    
+    void stopActiveStoppables() {
+        std::unordered_set<CBLStoppable*> stoppables;
+        {
+            std::unique_lock<std::mutex> lock(_stopMutex);
+            if (_stopping)
+                return;
+            _stopping = true;
+            stoppables = _stoppables;
+        }
+        
+        // Call stop outside lock to prevent deadlock:
+        for (auto s : stoppables) {
+            s->stop();
+        }
+        
+        {
+            std::unique_lock<std::mutex> lock(_stopMutex);
+            if (!_stoppables.empty()) {
+                CBL_Log(kCBLLogDomainDatabase, kCBLLogInfo,
+                        "Waiting for %zu active replicators and live queries to stop ...",
+                        _stoppables.size());
+                _stopCond.wait(lock, [this] {return _stoppables.empty();});
+            }
+        }
+    }
+    
+    bool registerStoppable(CBLStoppable* stoppable) {
+        LOCK(_stopMutex);
+        if (_stopping)
+            return false;
+        _stoppables.insert(stoppable);
```

Comment:
Is this object guaranteed to outlive the database?  It seems like it might be safer to retain the `stoppable` here and then release it in the `unregisterStoppable` function.

Reply:
First I thought, it is a good idea to make it safe but when the current way of removing the listener (token), I cannot retain. Also I don't want to touch those area at this point in time because it is more complicated than expected.

Anyway, to answer your question, it's pretty much guaranteed. There is one case that it could fail is when the replicator is released while it's still running. If that is the case, I would expect some problems including crash will occur before the database is closed.